In [1]:
import os
import pandas as pd

In [2]:
os.getcwd()

'/Users/lizzysterling'

In [3]:
path = "/Users/lizzysterling/Desktop/Term4/Capstone/PostsecondaryAttainment/RawDatasets"

In [4]:
os.chdir(path)

### We want to change all numeric columns to type numeric so that we can impute the mean, so we need to check the data types

In [5]:
d = pd.read_excel("mergedData.xlsx")

### Check to see how many null values are in each column

In [ ]:
d.isnull().sum()

#### Check the data types

In [ ]:
d.dtypes

#### We can see that there are many columns that are meant to be floats that are listed as objects. We need to iteratively go through all of the columns and change these values

In [6]:
for column in range(len(d.columns)): #loops through every column in the data set
    if d.iloc[:,column].isnull().sum() == 0: #these are the string columns that do not need to be changed
        print "Column ", column , "remained the same"
    elif d.iloc[:,column].dtype == "object": #if the column is of type object (percentages)
        for value in range(len(d)):
            if pd.isnull(d.iloc[value,column]):
                continue
            elif "%" in d.iloc[value,column]: # one of the issues is percentages
                d.iloc[value,column] = float(d.iloc[value,column].strip("%"))/100
            elif "," in d.iloc[value, column]: # another issues is commas
                d.iloc[value, column] = float(d.iloc[value, column].replace(",", ""))
            elif d.iloc[value,column] == "*":
                d.iloc[value, column] = None
            else:
                continue
        print "Column ", column, "was changed to type numeric"
        d.iloc[:,column] = pd.to_numeric(d.iloc[:,column])
    else:
        print "Column", column, "stayed the same"
                
       
    
                

Column  0 remained the same
Column  1 remained the same
Column  2 was changed to type numeric
Column 3 stayed the same
Column 4 stayed the same
Column  5 was changed to type numeric
Column 6 stayed the same
Column  7 was changed to type numeric
Column  8 was changed to type numeric
Column  9 was changed to type numeric
Column  10 was changed to type numeric
Column  11 was changed to type numeric
Column  12 was changed to type numeric
Column 13 stayed the same
Column  14 was changed to type numeric
Column 15 stayed the same
Column  16 was changed to type numeric
Column 17 stayed the same
Column  18 was changed to type numeric
Column 19 stayed the same
Column  20 was changed to type numeric
Column 21 stayed the same
Column  22 was changed to type numeric
Column 23 stayed the same
Column  24 was changed to type numeric
Column  25 was changed to type numeric
Column  26 was changed to type numeric
Column  27 was changed to type numeric
Column 28 stayed the same
Column  29 was changed to typ

ValueError: invalid literal for float(): 86% 100

In [ ]:
d.dtypes

It appears that all of our previous object types are now floats, so we can now impute using the mean values.

*Note that the first two columns are still objects - they contain string values indicating county and college name*

In [7]:
df = d.fillna(d.mean()) # run this once all of the columns are of the proper type

In [8]:
df.isnull().sum()

County                                                              0
College Name                                                        0
AdvESL_MeasureableSkills_Participant_POP_MSG                        0
AdvESL_MeasureableSkills_IndividualsServed                          0
AdvESL_MeasureableSkills_Participants12+Hours                       0
AdvESL_MeasureableSkills_ParticipServed                             0
AdvESL_MeasureableSkills_POPs                                       0
AdvESL_MeasureableSkills_AHSGrad                                    0
AdvESL_MeasureableSkills_HSE                                        0
AdvESL_MeasureableSkills_Postsecondary\rEnrollment                  0
AdvESL_MeasureableSkills_AHSCredits                                 0
AdvESL_MeasureableSkills_Post-test                                  0
AdvESL_MeasureableSkills_MSG                                        0
Beg_ESL_Students                                                    0
Beg_ESL_PCTProgress 

We can see that many of our columns are now null-free, but we need to double check since we have so many columns

In [9]:
stillNA = df.columns[df.isna().any()].tolist()
len(stillNA)

149

We have 149 columns that still contain null values. For the sake of analysis, we are going to remove these columns, but we can come back to them if we need to

In [10]:
type(stillNA)

list

In [22]:
for name in stillNA:
    #print name
    df = df.drop(name, axis=1)
    

Check that we succesfully removed the columns that contain NAs

In [23]:
df.columns[df.isna().any()].tolist()

#df1 = df.drop("NucMedTech_PCTPassing2015", axis = 0)

[]

In [26]:
newPath = "/Users/lizzysterling/Desktop/Term4/Capstone/PostsecondaryAttainment/RawDatasets"

In [27]:
os.chdir(newPath)

In [28]:
df.to_excel('imputedData.xlsx', index = False)

In [ ]:
finalData.to_excel('mergedData.xlsx', index=False)